cpp/include/cudf/detail/scatter.cuh

```diff
@@ -0,0 +1,193 @@
+
+    // NOTE use source.begin + scatter rows rather than source.end in case the
+    // scatter map is smaller than the number of source rows
+    thrust::scatter(rmm::exec_policy(stream)->on(stream), source.begin<T>(),
+      source.begin<T>() + std::distance(scatter_map_begin, scatter_map_end), 
+      scatter_map_begin,
+      result_view.begin<T>());
+
+    return result;
+  }
+
+  template <typename T, std::enable_if_t<not is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      MapIterator scatter_map_begin, MapIterator scatter_map_end, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    using strings::detail::create_string_vector_from_column;
+    auto const source_vector = create_string_vector_from_column(source, stream);
+    auto const begin = source_vector.begin();
+    auto const end = begin + std::distance(scatter_map_begin, scatter_map_end);
+    return strings::detail::scatter(begin, end, scatter_map_begin, target, mr, stream);
+  }
+};
+
+} //namespace
+
+/**
+ * @brief Scatters the rows of the source table into a copy of the target table
+ * according to a scatter map.
+ *
+ * Scatters values from the source table into the target table out-of-place,
+ * returning a "destination table". The scatter is performed according to a
+ * scatter map such that row `scatter_begin[i]` of the destination table gets row
+ * `i` of the source table. All other rows of the destination table equal
+ * corresponding rows of the target table.
+ *
+ * The number of columns in source must match the number of columns in target
+ * and their corresponding datatypes must be the same.
+ *
+ * If the same index appears more than once in the scatter map, the result is
+ * undefined.
+ *
+ * @throws cudf::logic_error if scatter_map.size() > source.num_rows()
+ *
+ * @param[in] source The input columns containing values to be scattered into the
+ * target columns
+ * @param[in] scatter_map_begin Beginning of iterator range of integer indices that map the rows in the
+ * source columns to rows in the target columns
+ * @param[in] scatter_map_end End of iterator range of integer indices that map the rows in the
+ * source columns to rows in the target columns
+ * @param[in] target The set of columns into which values from the source_table
+ * are to be scattered
+ * @param[in] check_bounds Optionally perform bounds checking on the values of
+ * `scatter_map` and throw an error if any of its values are out of bounds.
+ * @param[in] mr The resource to use for all allocations
+ * @param[in] stream The stream to use for CUDA operations
+ *
+ * @return Result of scattering values from source to target
+ **/
+template <typename T, typename MapIterator>
+std::unique_ptr<table>
+scatter(table_view const& source, MapIterator scatter_map_begin,
+    MapIterator scatter_map_end, table_view const& target,
+    bool check_bounds = false,
```

Comment:
This doesn't do anything. The API shouldn't accept this argument if it does nothing with it. Alternatively, you can pull out the bounds checking code, make it a separate function and call it from inside this one.